### PR TITLE
ROS Video Plugin for Gazebo - allows displaying an image stream in an OG...

### PR DESCRIPTION
**Pull request by piyushk**
...RE texture inside gazebo. Also provides a fix for #85.

### CODE REVIEW

File: gazebo_plugins/src/gazebo_ros_video.cc [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    std::string name = robot_namespace_ + "_visual";
+    video_visual_.reset(
+        new VideoVisual(name, parent, height, width));
+
+    // Initialize the ROS node and subscribe to cmd_vel
+    int argc = 0;
+    char** argv = NULL;
+    ros::init(argc, argv, "gazebo_ros_video_plugin",
```
Comment (davetcoleman) [inline gazebo_plugins/src/gazebo_ros_video.cc]: You should not initialize your own ROS node. See the example [here](http://gazebosim.org/wiki/Tutorials/1.9/Creating_ROS_plugins_for_Gazebo)

**Comment by davetcoleman** [comment]
Looks good otherwise, thanks for adding this!

**Comment by piyushk** [comment]
Removed ros initialization. Thanks for the pointer!

**Comment by piyushk** [comment]
Hmmm. I am facing a couple of problems with this:
- I am failing the ros::isInitialized check in both the constructor and Load function of the plugin. I am not sure why, as gazebo_ros_api_plugin is already available after running gzserver
- Should the check for ros initialization be in the constructor or the Load function?

**Comment by davetcoleman** [comment]
It should be in the Load function, the example on the tutorial was wrong, my bad!

File: gazebo_plugins/src/gazebo_ros_video.cc [hunk 2]
```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2013 Open Source Robotics Foundation
```
Comment (davetcoleman) [inline gazebo_plugins/src/gazebo_ros_video.cc]: Can you change the file name to .cpp to match everything else?

File: gazebo_plugins/src/gazebo_ros_video.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+#include <gazebo_plugins/gazebo_ros_video.h>
+#include <boost/lexical_cast.hpp>
+
+namespace gazebo {
+
+  VideoVisual::VideoVisual(
+      const std::string &name, rendering::VisualPtr parent, 
+      int height, int width) : 
+      rendering::Visual(name, parent), height_(height), width_(width) {
```
Comment (davetcoleman) [inline gazebo_plugins/src/gazebo_ros_video.cc]: And can you move the '{' to the next line on your functions please?

**Comment by piyushk** [comment]
I'll make the style changes. I am debugging a weird error with the plugin that I found after I removed the initialization call - it seems that Load for this plugin is getting called twice:
- One of the Loads is successful:

```
[ INFO] [1375119335.239259291, 0.629000000]: GazeboRosVideo (ns = /robot1/) has started!
```
- The other one is not:

```
[FATAL] [1375119334.310344044]: GazeboRosVideo Plugin (ns = /robot1/). A ROS node for Gazebo has not been initialized, unable to load plugin. Load the Gazebo system plugin 'libgazebo_ros_api_plugin.so' in the gazebo_ros package)
```

I'll update the pull request later today once I have debugged this error.

**Comment by davetcoleman** [comment]
I talked to John just now about this and we tested your code. The problem is that your visual rendering plugin is running for both gzserver and gzclient, which are two different processes. The gzserver already has a ros node initialized, but the gzclient does not. This is the first plugin in gazebo_plugins to be a Visual type, so I didn't realized this would happen but my advice to you to remove the ros::init() was wrong - add that back in :) Maybe name the node "gazebo_client" to indicate what process it belongs to?

**Comment by piyushk** [comment]
Thanks for the explanation, as I now understand all my observations.

I have updated the code with the style changes. A node is only initialized if necessary, and there is a small hack which attempts to print whether the instance belongs to gzserver or gzclient. It will have to do for now.

The PR is good to merge on my part, let me know if you spot anything amiss.

**Comment by davetcoleman** [comment]
I got it fully working on Baxter just now, I'm pretty excited about this new plugin as Rethink will probably us it!

As a side note from @hsu, your implementation as I understand it does not actually write to the textures, it only displays the image in gzclient. Meaning if a camera or a video is made of the simulation, the video stream will not appear. A better future implementation would actually change the link's texture. But this is still really great work!

Could you please document this plugin on the [plugins wiki](http://gazebosim.org/wiki/Tutorials/1.9/ROS_Motor_and_Sensor_Plugins) and include an XML example of the necessary gazebo tags? Thanks!

**Comment by piyushk** [comment]
I'll document the plugin. I've got another plugin I want to submit a PR for.

@hsu That's odd. I don't modify the link texture, but the external texture is visible in a camera/video stream (That's what I use it for). See http://farnsworth.csres.utexas.edu/share/gazebo_ros_video_output.png (I've put an output from my webcam on the simulated laptop screen, which is viewed by the camera on the controllable person model).

**Comment by davetcoleman** [comment]
bump to document the plugin on the wiki

**Pull request merged**